ipam/ring/ring.go

```diff
@@ -292,17 +303,38 @@ func (es entries) merge(other entries, ourPeer mesh.PeerName) (result entries, u
 			switch {
 			case mine.Version >= theirs.Version:
 				if mine.Version == theirs.Version && !mine.Equal(theirs) {
-					err = errInconsistentEntry(mine, theirs)
-					return
+					if mine.Peer == ourPeer {
+						// if we own the entry and has allocations
+						if checkEntryHasAllocations(mine, es.entry(i+1)) {
+							err = errInconsistentEntry(mine, theirs)
+							return
+						}
+					}
+					// tie-break here, pick the entry with the highest free count
+					if mine.Free >= theirs.Free {
+						addToResult(*mine)
+						previousOwner = &mine.Peer
+					} else {
+						addTheirs(*theirs)
+					}
+				} else {
+					addToResult(*mine)
+					previousOwner = &mine.Peer
 				}
-				addToResult(*mine)
-				previousOwner = &mine.Peer
 			case mine.Version < theirs.Version:
-				if mine.Peer == ourPeer { // We shouldn't receive updates to our own tokens
-					err = errNewerVersion(mine, theirs)
-					return
+				if mine.Peer == ourPeer {
+					// We received update to our own tokens accept the received entry
+					// if either it belongs to a different peer and we do not have allocations
+					// in the range effectively given away, or it belongs to our own peer
+					if theirs.Peer != ourPeer && !checkEntryHasAllocations(mine, es.entry(i+1)) || theirs.Peer == ourPeer {
+						addTheirs(*theirs)
```

Comment:
This is not implementing what it says in #1962 for the latter case:
>  we should set our version to the one received plus one, effectively imposing our existing entry.

Reply:
I tried the suggestion earlier, but was failing for this case (not sure if this in-valid test scenario)

```
	// we receive an entry that splits one of our ranges, giving some of it
	// away to another peer. accept the entry provided that we have no allocations
	// in the range that got given away.
	ring1 = NewRing(start, end, peer1name)
	ring2 = NewRing(start, end, peer2name)
	ring1.Entries = []*entry{{Token: start, Peer: peer1name, Free: 128, Version: 1}, {Token: middle, Peer: peer2name, Free: 128}}
	ring2.Entries = []*entry{{Token: start, Peer: peer1name, Free: 8, Version: 2}, {Token: dot8, Peer: peer2name, Free: 120, Version: 1}, {Token: middle, Peer: peer2name, Free: 128}}
	require.NoError(t, merge(ring1, ring2))
```

` imposing our existing entry` would mean keeping the free count as it is.

Reply:
The `Allocator` calls `reportFreeSpace()` after every action in `actorLoop()`, which should correct the free space.
I can't immediately see why that test case fails.

Reply:
After lengthy discussion, I understand the test fails because `Merge()` checks validity of everything before returning.

There is another case where accepting the incoming data is incorrect: if we have allocated some IPs in the free space before the update arrives.

So, I'm thinking that recalculating the free space when accepting an unexpected update is the way to go.

Reply:
> recalculating the free space when accepting an unexpected update is the way to go
`ring` does not have access to `space` to re-calculate the free space (need invasive changes across allocator, ring, space and tests to accommodate this). Given the conflicting conditions (merging of un-allocated ranges) this PR is addressing, I made change to reset the `Free` to distance between the tokens. PTAL